File: android/app/src/main/java/com/rajarsheechatterjee/ZipArchive/ZipArchive.java
package com.rajarsheechatterjee.ZipArchive;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.Map;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;
import java.util.zip.ZipOutputStream;

public class ZipArchive extends ReactContextBaseJavaModule {
    ZipArchive(ReactApplicationContext context) throws Exception {
        super(context);
    }

    @NonNull
    @Override
    public String getName() {
        return "ZipArchive";
    }

    private String escapeFilePath(String filePath){
        return filePath.replaceAll(":", "\uA789");
    }

    private void unzipProcess(ZipInputStream zis, String distDirPath) throws Exception {
        ZipEntry zipEntry;
        int len;
        byte[] buffer = new byte[4096];
        while ((zipEntry = zis.getNextEntry()) != null) {
            if(zipEntry.getName().endsWith("/")) continue;
            String escapedFilePath = this.escapeFilePath(zipEntry.getName());
            File newFile = new File(distDirPath, escapedFilePath);
            newFile.getParentFile().mkdirs();
            FileOutputStream fos = new FileOutputStream(newFile);
            while ((len = zis.read(buffer)) > 0) fos.write(buffer, 0, len);
            fos.close();
        }
        zis.closeEntry();
    }
    @ReactMethod
    public void unzip(String sourceFilePath, String distDirPath, Promise promise) {
        new Thread(new Runnable() {
            @Override
            public void run() {
                try{
                    ZipInputStream zis = new ZipInputStream(new FileInputStream(sourceFilePath));
                    unzipProcess(zis, distDirPath);
                    zis.close();
                    promise.resolve(null);
                } catch (Exception e) {
                    promise.reject(e.getCause());
                }
            }
        }).start();
    }

    @ReactMethod
    public void remoteUnzip(String distDirPath, String _url, @Nullable ReadableMap headers, Promise promise){
        new Thread(new Runnable() {
            @Override
            public void run() {
                try {
                    URL url = new URL(_url);
                    HttpURLConnection connection = (HttpURLConnection) url.openConnection();
                    connection.setRequestMethod("GET");
                    if(headers != null){
                        Iterator<Map.Entry<String, Object>> it = headers.getEntryIterator();
                        while (it.hasNext()){
                            Map.Entry<String, Object> entry = it.next();
                            connection.setRequestProperty(entry.getKey(), entry.getValue().toString());
                        }
                    }
                    ZipInputStream zis = new ZipInputStream(connection.getInputStream());
                    unzipProcess(zis, distDirPath);
                    connection.disconnect();
                    promise.resolve(null);
                }catch (Exception e){
                    promise.reject(e.getCause());
                }
            }
        }).start();
    }

    private ArrayList<String> walkDir(String path) {
        ArrayList<String> res = new ArrayList<>();
        File node = new File(path);
        if(node.isFile()){
            res.add(path);
        }else{
            String[] children = node.list();
            if(children != null){
                for(String filename: children){
                    ArrayList<String> childPaths = walkDir(new File(path, filename).toString());
                    res.addAll(childPaths);
                }
            }
        }
        return res;
    }
    private void zipProcess(String sourceDirPath, ZipOutputStream zos) throws Exception{
        ArrayList<String> paths = walkDir(sourceDirPath);
        byte[] buffer = new byte[4096];
        int len;
        for(String path: paths){
            ZipEntry zipEntry = new ZipEntry(path.replace(sourceDirPath + "/", ""));
            zos.putNextEntry(zipEntry);
            try (FileInputStream fis = new FileInputStream(path)) {
                while ((len = fis.read(buffer)) > 0) {
                    zos.write(buffer, 0, len);
                }
            }
            zos.closeEntry();
        }
        zos.close();
    }
    @ReactMethod
    public void remoteZip(String sourceDirPath, String _url, @Nullable ReadableMap headers, Promise promise){
        new Thread(new Runnable() {
            @Override
            public void run() {
                try {
                    URL url = new URL(_url);
                    HttpURLConnection connection = (HttpURLConnection) url.openConnection();
                    connection.setRequestMethod("POST");
                    if(headers != null){
                        Iterator<Map.Entry<String, Object>> it = headers.getEntryIterator();
                        while (it.hasNext()){
                            Map.Entry<String, Object> entry = it.next();
                            connection.setRequestProperty(entry.getKey(), entry.getValue().toString());
                        }
                    }
                    ZipOutputStream zos = new ZipOutputStream(connection.getOutputStream());
                    zipProcess(sourceDirPath, zos);
                    if(connection.getResponseCode() == 200){
                        InputStream is = connection.getInputStream();
                        ByteArrayOutputStream result = new ByteArrayOutputStream();
                        byte[] buffer = new byte[1024];
                        for (int length; (length = is.read(buffer)) != -1; ) {
                            result.write(buffer, 0, length);
                        }
                        is.close();
                        connection.disconnect();
                        promise.resolve(result.toString());
                    }else{
                        throw new Exception("Network request failed");
                    }
                }catch (Exception e){
                    promise.reject(e.getCause());
                }
            }
        }).start();
    }
}

File: android/app/src/main/java/com/rajarsheechatterjee/ZipArchive/ZipArchive.kt
package com.rajarsheechatterjee.ZipArchive

import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import java.io.File
import java.io.FileOutputStream
import java.net.HttpURLConnection
import java.net.URL
import java.util.zip.ZipEntry
import java.util.zip.ZipFile
import java.util.zip.ZipInputStream
import java.util.zip.ZipOutputStream

class ZipArchive(context: ReactApplicationContext) : ReactContextBaseJavaModule(context) {
    override fun getName(): String {
        return "ZipArchive"
    }

    @ReactMethod
    fun unzip(sourceFilePath: String, distDirPath: String, promise: Promise) {
        Thread {
            try {
                ZipFile(sourceFilePath).use { zis ->
                    zis.entries().asSequence().filterNot { it.isDirectory }.forEach { zipEntry ->
                        val newFile = File(distDirPath, zipEntry.name)
                        newFile.parentFile?.mkdirs()
                        zis.getInputStream(zipEntry).use { inputStream ->
                            FileOutputStream(newFile).use { fos -> inputStream.copyTo(fos, 4096) }
                        }
                    }
                }
                promise.resolve(null)
            } catch (e: Exception) {
                promise.reject(e)
            }
        }.start()
    }

    @ReactMethod
    fun remoteUnzip(
        distDirPath: String,
        urlString: String,
        headers: ReadableMap,
        promise: Promise
    ) {
        val connection = URL(urlString).openConnection() as HttpURLConnection
        Thread {
            try {
                connection.requestMethod = "GET"
                val it = headers.entryIterator
                while (it.hasNext()) {
                    val (key, value) = it.next()
                    connection.setRequestProperty(key, value.toString())
                }
                ZipInputStream(connection.inputStream).use { zis ->
                    generateSequence { zis.nextEntry }
                        .filterNot { it.isDirectory }
                        .forEach { zipEntry ->
                            val newFile = File(distDirPath, zipEntry.name)
                            newFile.parentFile?.mkdirs()
                            FileOutputStream(newFile).use { fos -> zis.copyTo(fos, 4096) }
                        }
                }
                if (connection.responseCode == 200) {
                    promise.resolve(null)
                } else {
                    throw Exception("Network request failed")
                }
            } catch (e: Exception) {
                promise.reject(e)
            } finally {
                connection.disconnect()
            }
        }.start()
    }

    private fun zipProcess(sourceDirPath: String, zos: ZipOutputStream) {
        val sourceDir = File(sourceDirPath)
        sourceDir.walkBottomUp().filter { it.isFile }.forEach { file ->
            val zipFileName =
                file.absolutePath.removePrefix(sourceDir.absolutePath).removePrefix("/")
            val entry = ZipEntry("$zipFileName${(if (file.isDirectory) "/" else "")}")
            zos.putNextEntry(entry)
            file.inputStream().use { fis -> fis.copyTo(zos, 4096) }
        }
    }

    @ReactMethod
    fun remoteZip(
        sourceDirPath: String,
        urlString: String,
        headers: ReadableMap,
        promise: Promise
    ) {
        Thread {
            val connection = URL(urlString).openConnection() as HttpURLConnection
            try {
                connection.requestMethod = "POST"
                val it = headers.entryIterator
                while (it.hasNext()) {
                    val (key, value) = it.next()
                    connection.setRequestProperty(key, value.toString())
                }
                ZipOutputStream(connection.outputStream).use { zipProcess(sourceDirPath, it) }
                if (connection.responseCode == 200) {
                    promise.resolve(
                        connection.inputStream.bufferedReader().use { it.readText() })
                } else {
                    throw Exception("Network request failed")
                }
            } catch (e: Exception) {
                promise.reject(e)
            } finally {
                connection.disconnect()
            }
        }.start()
    }
}
